traitsui/tests/test_undo.py

```diff
@@ -25,6 +27,1075 @@ class GuiTestAssistant(object):
+            new_value=('foo', 'wombat', 'baz', 'fizz'),
+        )
+
+        result = undo_item.merge(next_undo_item)
+
+        self.assertFalse(result)
+
+    def test_merge_unhandled_type(self):
+        example = SimpleExample()
+
+        undo_item = UndoItem(
+            object=example,
+            name='any_value',
+            old_value={'foo', 'bar', 'baz'},
+            new_value={'foo', 'wombat', 'baz'},
+        )
+        next_undo_item = UndoItem(
+            object=example,
+            name='any_value',
+            old_value={'foo', 'wombat', 'baz'},
+            new_value={'foo', 'fizz', 'baz'},
+        )
+
+        result = undo_item.merge(next_undo_item)
+
+        self.assertFalse(result)
+
+
+class TestListUndoItem(UnittestTools, unittest.TestCase):
+
+    def test_undo(self):
+        example = SimpleExample(list_value=['foo', 'wombat', 'baz'])
+
+        undo_item = ListUndoItem(
+            object=example,
+            name='list_value',
+            index=1,
+            added=['wombat', 'baz'],
+            removed=['bar'],
+        )
+
+        with self.assertTraitChanges(example, 'list_value_items', count=1):
+            undo_item.undo()
+
+        self.assertEqual(example.list_value, ['foo', 'bar'])
+
+    def test_redo(self):
+        example = SimpleExample(list_value=['foo', 'bar'])
+
+        undo_item = ListUndoItem(
+            object=example,
+            name='list_value',
+            index=1,
+            added=['wombat', 'baz'],
+            removed=['bar'],
+        )
+
+        with self.assertTraitChanges(example, 'list_value_items', count=1):
+            undo_item.redo()
+
+        self.assertEqual(example.list_value, ['foo', 'wombat', 'baz'])
+
+    def test_merge_identical(self):
+        example = SimpleExample(list_value=['foo', 'wombat', 'baz'])
+        added = [['wombat'], 'baz']
+        removed = [['bar']]
+
+        undo_item = ListUndoItem(
+            object=example,
+            name='list_value',
+            index=1,
+            added=added.copy(),
+            removed=removed.copy(),
+        )
+        next_undo_item = ListUndoItem(
+            object=example,
+            name='list_value',
+            index=1,
+            added=added.copy(),
+            removed=removed.copy(),
+        )
+
+        result = undo_item.merge(next_undo_item)
+
+        self.assertTrue(result)
+
+    def test_merge_equal(self):
+        example = SimpleExample(list_value=['foo', 'wombat', 'baz'])
+        removed = [['bar']]
+
+        undo_item = ListUndoItem(
+            object=example,
+            name='list_value',
+            index=1,
+            added=[['wombat'], 'baz'],
+            removed=removed.copy()
+        )
+        next_undo_item = ListUndoItem(
+            object=example,
+            name='list_value',
+            index=1,
+            added=[['wombat'], 'baz'],
+            removed=removed.copy(),
+        )
+
+        result = undo_item.merge(next_undo_item)
```

Comment:
This is testing declared behaviour from this comment:
https://github.com/enthought/traitsui/blob/79b66f5670e3de2b73529db5c29b3d216f0a2389/traitsui/undo.py#L242-L244

Since it is declared it should be tested that the code does what it says; if we feel that relaxing to equality is desirable, we should modify the code and the comment, but I don't have enough insight into whether that would be a problem.